source/common/config/version_converter.cc

```diff
@@ -40,5 +92,59 @@ DowngradedMessagePtr VersionConverter::downgrade(const Protobuf::Message& messag
   return downgraded_message;
 }
 
+DynamicMessagePtr
+VersionConverter::reinterpret(const Protobuf::Message& message,
+                              envoy::config::core::v3alpha::ApiVersion api_version) {
+  switch (api_version) {
+  case envoy::config::core::v3alpha::ApiVersion::AUTO:
+  case envoy::config::core::v3alpha::ApiVersion::V2:
+    // TODO(htuch): this works as long as there are no new fields in the v3+
+    // DiscoveryRequest. When they are added, we need to do a full v2 conversion
+    // and also discard unknown fields. Tracked at
+    // https://github.com/envoyproxy/envoy/issues/9619.
+    return downgrade(message);
+  case envoy::config::core::v3alpha::ApiVersion::V3ALPHA: {
+    // We need to scrub the hidden fields.
+    auto non_shadowed_message = std::make_unique<DynamicMessage>();
+    const Protobuf::Descriptor* desc = message.GetDescriptor();
+    non_shadowed_message->msg_.reset(
+        non_shadowed_message->dynamic_msg_factory_.GetPrototype(desc)->New());
+    non_shadowed_message->msg_->MergeFrom(message);
+    VersionUtil::scrubHiddenEnvoyDeprecated(*non_shadowed_message->msg_);
+    return non_shadowed_message;
+  }
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+bool VersionUtil::hasHiddenEnvoyDeprecated(const Protobuf::Message& message) {
+  bool has_hidden = false;
+  constMessageFieldVisitor(
+      [&has_hidden](const Protobuf::Message& message, const Protobuf::FieldDescriptor& field,
+                    const Protobuf::Reflection& reflection) {
+        if (absl::StartsWith(field.name(), "hidden_envoy_deprecated_")) {
```

Comment:
nit: store "hidden_envoy_deprecated_" in some const somewhere and use all the places its needed?